### PR TITLE
Create a checkpoint only at the end of Compact call

### PR DESCRIPTION
**Pull request by codesome**
Fixes https://github.com/prometheus/prometheus/issues/7574 https://github.com/prometheus/prometheus/issues/6359

### CODE REVIEW

**Comment by codesome** [comment]
This PR moves the checkpointing to the end of all head compaction in the loop instead of each head compaction.

(If you saw this PR before this comments, it was truncating the head at the end, but that has problems as we need to advance the mint of the head. So, now, we are only moving the checkpointing to the end)

**Comment by beorn7** [comment]
This looks good to me as only being semi-qualified in TSDB matters.

@bwplotka @krasi-georgiev would be great to get this reviewed today so that I can include it into the 2.20 RC today.

**Comment by codesome** [comment]
@beorn7 if this PR gets delayed, feel free to go ahead with 2.20 RC as this is merely an enhancement which can go later too.

**Comment by pstibrany** [comment]
Previously `prometheus_tsdb_wal_truncate_duration_seconds` covered checkpoint time as well. After this change, it will not. I think we need new metric for checkpoint times.

**Comment by codesome** [comment]
@pstibrany It still does. What I have done is just split `Truncate()` into `Truncate()` and `performCheckpoint()`, and the `prometheus_tsdb_wal_truncate_duration_seconds` calculation covers the entire `performCheckpoint()` (WAL truncation and checkpoint creation)

**Comment by pstibrany** [comment]
> @pstibrany It still does. What I have done is just split `Truncate()` into `Truncate()` and `performCheckpoint()`, and the `prometheus_tsdb_wal_truncate_duration_seconds` calculation covers the entire `performCheckpoint()` (WAL truncation and checkpoint creation)

Ah, right. I was thinking about previous version of the code. Sorry and thanks for correcting me.

File: tsdb/db.go [hunk 1]
```diff
@@ -740,6 +740,16 @@ func (db *DB) Compact() (err error) {
 	}()
 	// Check whether we have pending head blocks that are ready to be persisted.
 	// They have the highest priority.
+	checkpointTime := int64(math.MinInt64)
+	checkpointAttempted := false
+	defer func() {
+		if err != nil && !checkpointAttempted {
```
Comment (pracucci) [inline tsdb/db.go:746]: If `db.stopc` is closed, no checkpoint is performed because of the check `err != nil`. Is it intended? Previous code would have run it.

File: tsdb/db.go [hunk 2]
```diff
@@ -760,52 +770,82 @@ func (db *DB) Compact() (err error) {
...
+		t, err := db.compactHead(head)
+		if err != nil {
 			return err
 		}
+		if t > checkpointTime {
+			checkpointTime = t
+		}
+	}
+
+	checkpointAttempted = true
```
Comment (pracucci) [inline tsdb/db.go:782]: If head was not compactable at first `for` iteration, is it correct we try to perform checkpoint?

File: tsdb/db.go [hunk 3]
```diff
@@ -760,52 +770,82 @@ func (db *DB) Compact() (err error) {
...
 
+func (db *DB) performHeadCheckpoint(t int64) error {
+	if blocks := db.Blocks(); len(blocks) > 0 {
+		blockMaxt := blocks[len(blocks)-1].Meta().MaxTime
+		if blockMaxt > t {
+			t = blockMaxt
+		}
+	}
+
+	if t == math.MinInt64 {
```
Comment (pracucci) [inline tsdb/db.go:798]: Uhm, zero value of `t` is `0`, not `math.MinInt64`. I think you have to initialise `t = math.MinInt64` at the beginning of the function, no?

Comment (roidelapluie) [reply tsdb/db.go:798]: t is a parameter

File: tsdb/db.go (hunk 1 above)
Comment (codesome) [reply tsdb/db.go:746]: Good point! We need to truncate.

File: tsdb/db.go (hunk 2 above)
Comment (codesome) [reply tsdb/db.go:782]: Right, we should not :) thanks for pointing out

File: tsdb/db.go (hunk 3 above)
Comment (codesome) [reply tsdb/db.go:798]: `t` is an argument (hence should not initialize in the method), and we are using `math.MinInt64` as zero value (which comes from `compactHead` and `Compact`)

File: tsdb/db.go (hunk 1 above)
Comment (codesome) [reply tsdb/db.go:746]: s/truncate/checkpoint

File: tsdb/db.go (hunk 3 above)
Comment (pracucci) [reply tsdb/db.go:803]: 🤦

Comment (codesome) [reply tsdb/db.go:803]: As Golang's `time.Unix(x,y)` method accepts 0 and even negative numbers, I thought of keeping the zero value as minint64.

**Comment by codesome** [comment]
I have refactored a bit to keep the original behaviour of `Head.Truncate()` method (as it creates a lot of changes and confusion in unit tests), and instead introduced `Head.TruncateInMemory()` and `Head.TruncateWAL()` which is used during compaction here.

**Comment by roidelapluie** [comment]
Hello @bwplotka, can you please review the last changes here? Thanks!

**Comment by pstibrany** [comment]
> Have you considered adding few tests (only WAL no blocks no gap in the WAL, with gap in the WAL, blocks and WAL) and measuring that the expected number of checkpoints occurred?

Quoting this to make sure it's not forgotten ^ (it would increase the confidence in the change).

**Comment by codesome** [comment]
Finally found the bug causing the tests to fail after few hours of investigation, and the fix is a 1 liner :) https://github.com/prometheus/prometheus/pull/7579/commits/7c3dcba4760f474b433b487d80924a92b3298fd7

Adding more tests now.

**Comment by codesome** [comment]
PR is ready for another review. Should be much simpler to parse now, with added tests.

File: tsdb/head.go [hunk 4]
```diff
@@ -781,8 +781,21 @@ func (h *Head) removeCorruptedMmappedChunks(err error) map[uint64][]*mmappedChun
 	return mmappedChunks
 }
 
-// Truncate removes old data before mint from the head.
+// Truncate removes old data before mint from the head and WAL.
 func (h *Head) Truncate(mint int64) (err error) {
+	initialize := h.MinTime() == math.MaxInt64
+	dontTruncateWAL := initialize || (!initialize && h.MinTime() >= mint)
```
Comment (pracucci) [inline tsdb/head.go:782]:
```suggestion
	dontTruncateWAL := initialize || h.MinTime() >= mint
```

File: tsdb/db.go [hunk 5]
```diff
@@ -839,21 +884,22 @@ func (db *DB) compactHead(head *RangeHead) (err error) {
...
 	if (uid == ulid.ULID{}) {
-		// Compaction resulted in an empty block.
-		// Head truncating during db.reload() depends on the persisted blocks and
 		// in this case no new block will be persisted so manually truncate the head.
-		if err = db.head.Truncate(maxt); err != nil {
-			return errors.Wrap(err, "head truncate failed (in compact)")
+		// Head truncating during db.reload() depends on the persisted blocks and
+		// Compaction resulted in an empty block.
+		if err = db.head.TruncateInMemory(maxt); err != nil {
+			return walTruncationTime, errors.Wrap(err, "head truncate failed (in compact)")
```
Comment (pracucci) [inline tsdb/db.go:897]: This will return `walTruncationTime = maxt` but, unless I'm missing something, the returned `walTruncationTime` is never used if there's an error. I'm wondering if the could would be easier to follow if you always `return math.MinInt64` and at the end of this function you `return maxt, nil`

File: tsdb/db.go [hunk 6]
```diff
@@ -805,31 +816,65 @@ func (db *DB) Compact() (err error) {
...
+		if err != nil {
 			return err
 		}
+		if t > walTruncationTime {
+			walTruncationTime = t
+		}
+		headCompacted = true
+	}
+
+	if headCompacted {
```
Comment (pracucci) [inline tsdb/db.go:829]: What do you think to add a code comment to explain why we truncate the WAL only at the end?

**Review by pracucci — COMMENTED** [review]
Thanks @codesome for working on this improvement! To my limited knowledge of TSDB, changes make sense. The test also cover the 2 cases we've seen, so 👍 .

File: tsdb/db.go [hunk 7]
```diff
@@ -805,31 +816,65 @@ func (db *DB) Compact() (err error) {
...
+		walTruncationAttempted = true
+		if err := db.truncateWAL(walTruncationTime); err != nil {
+			return errors.Wrap(err, "WAL truncation in Compact")
```
Comment (bwplotka) [inline tsdb/db.go:832]: Let's add different wrap here vs in defer above (:

File: tsdb/db.go [hunk 8]
```diff
@@ -783,6 +783,17 @@ func (db *DB) Compact() (err error) {
 			db.metrics.compactionsFailed.Inc()
 		}
 	}()
+
+	walTruncationTime := int64(math.MinInt64)
+	walTruncationAttempted, headCompacted := false, false
+	defer func() {
+		if headCompacted && !walTruncationAttempted {
```
Comment (bwplotka) [inline tsdb/db.go:790]: Not a fan of this. This logic with truncating something in memory then here, then also when head is compacted is kind of hard to follow. Let me look if I can suggest something better quickly.

**Review by bwplotka — COMMENTED** [review]
Sorry for lag. Generally makes sense, but we might need slight refactor - the current proposal has many little flags and tweaks and might hard to follow. Ideally head totally mask WAL - it's optional part of head, so why DB has to manage it?

**Comment by bwplotka** [comment]
Created simplified version: https://github.com/prometheus/prometheus/pull/8067 on top of your PR. 

Diff vs this PR: https://github.com/codesome/prometheus/compare/truncate-once...prometheus:checkpoint-at-the-end

**Comment by codesome** [comment]
Closing this in favor of #8067, some minor thing to fix in that and it's LGTM for me :P
